File: software_architecture_model_python/src/software_architecture_model_python/tree/scanner.py
"""
Σάρωση directory και παραγωγή normalized tree structure.

Το module παρέχει την κλάση TreeScanner, η οποία:
- διαβάζει αναδρομικά όλους τους φακέλους
- δημιουργεί λεξικό parent → children
- χρησιμοποιείται ως input για τον ταξινομητή αρχιτεκτονικής

Αποτελεί το πρώτο στάδιο του pipeline αναγνώρισης αρχιτεκτονικής.
"""

from pathlib import Path


class TreeScanner:
    """
    Σκανάρει ένα directory και επιστρέφει ένα normalized tree structure.
    Το tree είναι dict: { "root": ["child1", "child2", ...], ... }
    """

    def scan(self, root_path: str) -> dict[str, list[str]]:
        """
    Σκανάρει ένα directory και επιστρέφει το tree structure.

    :param root_path: Το μονοπάτι του project.
    :type root_path: str
    :return: Λεξικό parent → children.
    :rtype: dict[str, list[str]]
    """
        root = Path(root_path)
        tree: dict[str, list[str]] = {}

        for path in root.rglob("*"):
            if path.is_dir():
                parent = str(path.parent)
                node = str(path)
                tree.setdefault(parent, []).append(node)

        return tree

File: js_package/src/tree/scanner.ts
/**
 * Σάρωση directory και παραγωγή normalized tree structure.
 *
 * Η κλάση TreeScanner:
 * - διαβάζει αναδρομικά όλους τους φακέλους
 * - δημιουργεί λεξικό parent → children
 * - χρησιμοποιείται ως input για τον ταξινομητή αρχιτεκτονικής
 *
 * Αποτελεί το πρώτο στάδιο του pipeline αναγνώρισης αρχιτεκτονικής.
 */

import { promises as fs } from "fs";
import path from "path";

export class TreeScanner {
  /**
   * Σκανάρει ένα directory και επιστρέφει το tree structure.
   *
   * Το tree είναι dict: { "root": ["child1", "child2", ...], ... }
   */
  async scan(rootPath: string): Promise<Record<string, string[]>> {
    const tree: Record<string, string[]> = {};

    async function walk(current: string) {
      const entries = await fs.readdir(current, { withFileTypes: true });

      for (const entry of entries) {
        if (entry.isDirectory()) {
          const fullPath = path.join(current, entry.name);
          const parent = current;
          const node = fullPath;

          if (!tree[parent]) {
            tree[parent] = [];
          }
          tree[parent].push(node);

          await walk(fullPath);
        }
      }
    }

    await walk(rootPath);
    return tree;
  }
}
